### PR TITLE
Preserve changes that are still relevant after revert of #66311

### PR DESCRIPTION
**Pull request by AlekseyTs**
Related to #63221.

### CODE REVIEW

**Review by 333fred — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@jcouv, @dotnet/roslyn-compiler For the second review

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs [hunk 1]
```diff
@@ -694,29 +694,23 @@ private void ReferToTempIfReferenceTypeReceiver(BoundLocal receiverTemp, ref Bou
...
                 assignmentToTemp =
                     assignmentToTemp.Update(
                         assignmentToTemp.Left,
-                        _factory.Conditional(
-                            isValueTypeCheck,
-                            intermediateRef,
-                            _factory.Sequence(new BoundExpression[] { _factory.AssignmentExpression(cache, intermediateRef) }, cache),
-                            receiverType,
-                            isRef: true),
+#pragma warning disable format
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:703]: nit: What's the formatting issue? Should we file a tracking issue?

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs [hunk 2]
```diff
@@ -564,6 +569,60 @@ private void EmitLoweredConditionalAccessExpression(BoundLoweredConditionalAcces
...
 
+        private sealed class IsConditionalConstrainedCallThatMustUseTempForReferenceTypeReceiverWalker : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+        {
+            private readonly BoundLoweredConditionalAccess _conditionalAccess;
+            private bool? _result;
+
+            private IsConditionalConstrainedCallThatMustUseTempForReferenceTypeReceiverWalker(BoundLoweredConditionalAccess conditionalAccess)
+                : base()
+            {
+                _conditionalAccess = conditionalAccess;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:580]: nit: it looks like the only part we need to hold onto is the `Id`

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:703]: > What's the formatting issue? Should we file a tracking issue?

IDE wants the initializer on a separate line. I think it is by design, but I do not like the design. I have no intention to fight for its change though.

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs [hunk 3]
```diff
@@ -564,6 +569,60 @@ private void EmitLoweredConditionalAccessExpression(BoundLoweredConditionalAcces
             }
         }
 
+        private sealed class IsConditionalConstrainedCallThatMustUseTempForReferenceTypeReceiverWalker : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+        {
+            private readonly BoundLoweredConditionalAccess _conditionalAccess;
+            private bool? _result;
+
+            private IsConditionalConstrainedCallThatMustUseTempForReferenceTypeReceiverWalker(BoundLoweredConditionalAccess conditionalAccess)
+                : base()
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:578]: nit: is the base initializer needed?

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs [hunk 4]
```diff
@@ -564,6 +569,60 @@ private void EmitLoweredConditionalAccessExpression(BoundLoweredConditionalAcces
             }
         }
 
+        private sealed class IsConditionalConstrainedCallThatMustUseTempForReferenceTypeReceiverWalker : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:572]: >IsConditionalConstrainedCallThatMustUseTempForReferenceTypeReceiverWalker [](http://example.com/codeflow?start=29&length=73)

nit: a comment would be nice "we must use a temp when ..."

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1) with some nits to consider

**Pull request merged**